### ISSUE TITLE
Make a Relay ConnectionField optimized to work with Query

### ISSUE THREAD

**Issue by Fedalto** [graphql-python/graphene-sqlalchemy#89]
Graphene has the `graphene.relay.ConnectionField`, which slices a list/iterable to build the edges inside the Relay connection.
But it doesn't play well with DB queries. [It calls `len()`](https://github.com/graphql-python/graphql-relay-py/blob/master/graphql_relay/connection/arrayconnection.py#L13) with `Query` doesn't have.

graphene-sqlalchemy should have a implementation of the ConnectionField that is aware of the sqlalchemy's Query and use it to slice directly in the DB.

For now, I'm using this workaround:

```py
class RelayConnectionField(graphene.relay.ConnectionField):
    @classmethod
    def resolve_connection(cls, connection_type, args, resolved):
        if isinstance(resolved, Query):
            len_ = resolved.count()
            connection = connection_from_list_slice(
                resolved,
                args,
                connection_type=connection_type,
                edge_type=connection_type.Edge,
                pageinfo_type=PageInfo,
                slice_start=0,
                list_length=len_,
                list_slice_length=len_,
            )
            connection.iterable = resolved
            return connection
        else:
            return super().resolve_connection(connection_type, args, resolved)
```

**Comment by Fedalto** [comment]
Similar/same issue to https://github.com/graphql-python/graphene/issues/589

**Comment by sebastiandev** [comment]
Isn't this already covered [here](https://github.com/graphql-python/graphene-sqlalchemy/blob/master/graphene_sqlalchemy/fields.py) ?

```python
    @classmethod
    def connection_resolver(cls, resolver, connection, model, root, info, **args):
        iterable = resolver(root, info, **args)
        if iterable is None:
            iterable = cls.get_query(model, info, **args)
        if isinstance(iterable, Query):
            _len = iterable.count()
        else:
            _len = len(iterable)
```

**Comment by Fedalto** [comment]
Indeed it is.
But if I used it I receives an error here (https://github.com/graphql-python/graphene-sqlalchemy/blob/master/graphene_sqlalchemy/fields.py#L26) as I'm sending a `graphene.relay.Connection` instead of a `SQLAlchemyObjectType`.
Maybe I'm using it incorrectly?

**Comment by sebastiandev** [comment]
Yeah you need a SQLAlchemyObjectType there, everything gets down to an ObjectType eventually
Show us your Query definition to have a better idea of whats going on

**Comment by Fedalto** [comment]
Following graphene 2.0 (https://github.com/graphql-python/graphene/blob/master/UPGRADE-v2.0.md#node-connections), I defined explicit Connections.
My code is very similar to the one described in the link:

```py
class User(SQLAlchemyObjectType):
    class Meta:
        model = UserModel
    name = String()

class UserConnection(relay.Connection):
    class Meta:
        node = User

class Group(SQLAlchemyObjectType):
    class Meta:
        model = GroupModel
    # Here, it works with the RelayConnectionField I posted in the issue.
    # But breaks if I change to SQLAlchemyConnectionField
    user_connection = RelayConnectionField(UserConnection)

class Query(ObjectType):
    group = Field(Group)
```

It does, however, work if I send the SQLAlchemyObjectType directly, like
```py
user_connection = SQLAlchemyConnectionField(User)
```
But how can I add custom data to the Connection itself?

**Comment by sebastiandev** [comment]
You are mixing RelayConnectionFields with SqlAlchemy,  You have to extend from SQLAlchemyConnectionField. I cant take a look right now, but check this as here we add fields to connections for counting purposes:
https://github.com/graphql-python/graphene-sqlalchemy/issues/58#issuecomment-349050744

**Comment by Fedalto** [comment]
@sebastiandev, thanks for the help.

`SQLAlchemyConnectionField` indeed does handle the `Query`.
My issue now is how to subclass a Connection to add custom data to it.
Seems a duplicate of #65 (and there's a PR to fix it already #94). 
I'm closing this.
